### ISSUE TITLE
[Refactor] IR standardization

### ISSUE THREAD

**Issue by yuanming-hu** [taichi-dev/taichi#689]
At some point, we should make the Taichi IR (a.k.a. CHI, CHI Hierarchical Instructions, 气) more standard. This will not only make IR more friendly to developers/more maintainable, but also allows others to reuse the IR to build new languages.

 - [ ] Unify `Stmt` fields (e.g. `AtomicOpStmt::dest/val` are inconsistent with `GlobalStoreStmt::ptr/data`)
 - [ ] `OffloadedStmt::loop_vars` and `OffloadedStmt::step` is not used (thanks for @xumingkuan)
 - [ ] Add namespace `analysis` (thanks for @xumingkuan)
 - [ ] Split frontend/backend IR. Everything in the frontend is unique to `taichi`, and the backend statements are reusable for build a different language.
 - [ ] Document IR

**Comment by k-ye** [comment]
Go CHI<s>吃</s> lang..!

**Comment by k-ye** [comment]
I wonder if we should separate `OffloadedStmt` out from the IR layer, and rename it to `OffloadedTask` or something. We will also have a special `Root` node representing the root of the AST, which contains a `std::vector<OffloadedTask>`. 

By doing so, we can embed more structural knowledge of a Taichi kernel into the types.  For example, a Taichi kernel is always composed of a series of `OffloadedTask`s (e.g. `{range_for, serial, range_for}, {clear, listgen, clear, listgen, struct_for, gc}`, ...). This is actually a characteristic that should be described by the types.

Right now the root is just a `Block`, and all its offloaded tasks are just inside `Block::statements`. This is no different from the blocks at the internal levels (e.g. `IfStmt`'s two blocks). So in our codegens, we have to keep track of whether we are handling root or not, with flags like `is_top_level_`.  This also makes it hard to look up info other than then current `OffloadedStmt`. E.g. when processing `struct_for`, I may want to look at the immediate previous `listgen` and `clear`.

WDYT?

**Comment by archibate** [comment]
> WDYT?

Can't agree more, also note that there are not nesting `OffloadStmt`, so let's just say `OffloadStmt` **is** root stmt, and a kernel contains multiple roots. We can do this by modifying:
https://github.com/taichi-dev/taichi/blob/62cdc60f2e3d981b4d5c1c9f60b7590c9e8031a1/taichi/program/kernel.h#L14
to something like:
```cpp
std::vector<OffloadStmt *> offloads;
```

**Comment by k-ye** [comment]
I believe having a `Root` is still worth the efforts. It is a proper encapsulation of the AST. As one example, think about the AST visitor interface -- it could be just `visit(Root&)` instead of `visit(std::vector<OffloadedTasks*>&)`. Having `Root` makes it a lot easier to extend without breaking the visitor's interface.

See also https://testing.googleblog.com/2017/11/obsessed-with-primitives.html

**Comment by archibate** [comment]
> `visit(std::vector<OffloadedTasks*>&)`

You mean, `class Root : Stmt`? Or `class Root { vector<OffloadedTasks> os; }`?

Also note that we want to implement real functions at some point #602, we want to leave a way for it when implementing `Root` or whatever.

**Comment by k-ye** [comment]
The later one.

> Also note that we want to implement real functions at some point #602, we want to leave a way for it when implementing Root or whatever.

Yeah, then it's even more reasonable that we have a `Root` to wrap up things like offloaded tasks, function defs, etc.

**Comment by xumingkuan** [comment]
Shall we put all functions in `analysis/` into `namespace analysis`?

**Comment by yuanming-hu** [comment]
Good idea. All passes that **do not change IR** should go to `irpass::analysis`.

**Comment by MarisaKirisame** [comment]
Is there gonna be serialization/pretty printer/parser so ppl can build external tool without interfacing C++?
